File: leet-code/371. Sum of Two Integers/SumOfTwoIntegers371_2.java
package com.example;

public class SumOfTwoIntegers371_2 {
    public int getSum(int a, int b) {
        while (b != 0) {
            // 자리올림 계산
            int carry = (a & b) << 1;
            // 자리합 계산 (carry 값을 고려하지 않음)
            a = a ^ b;
            // 자리올림을 b 에 할당
            b = carry;
        }
        return a;
    }
}

File: leet-code/371. Sum of Two Integers/SumOfTwoIntegers371_2.kt
package com.example

class SumOfTwoIntegers371_2 {
    fun getSum(a: Int, b: Int): Int {
        val binaryA = String.format("%32s", a.toUInt().toString(radix = 2)).replace(" ", "0")
        val binaryB = String.format("%32s", b.toUInt().toString(radix = 2)).replace(" ", "0")
        var sum = 0
        var carry = 0
        var answer = StringBuilder()
        for (i in 31 downTo 0) {
            val numA = binaryA[i].code - '0'.code
            val numB = binaryB[i].code - '0'.code
            sum = (numA xor numB) xor carry
            carry = ((numA xor numB) and carry) or (numA and numB)
            answer.insert(0, sum)
        }
        println(answer.toString())
        return Integer.parseUnsignedInt(answer.toString(), 2)
    }
}
